src/App.tsx

```diff
@@ -1,24 +1,289 @@
 
 export const App: React.FC = () => {
-  if (!USER_ID) {
-    return <UserWarning />;
-  }
+  const [todos, setTodos] = useState<TodoType[] | []>([]);
+  const [filter, setFilter] = useState<FilterType>(FilterType.All);
+  const [errorMessage, setErrorMessage] = useState('');
+  const [todoText, setTodoText] = useState('');
+  const [isInputDisabled, setIsInputDisabled] = useState(false);
+  const [tempTodo, setTempTodo] = useState<TodoType | null>(null);
+  const [loadingTodoIds, setLoadingTodoIds] = useState<number[]>([]);
+  const [editedTodoId, setEditedTodoId] = useState<number | null>(null);
+  const [editedTodoText, setEditedTodoText] = useState('');
+  const inputRef = useRef<HTMLInputElement>(null);
+
+  const handleLoadTodos = useCallback(async () => {
+    try {
+      setTodos(await getTodos(USER_ID));
+    } catch (error) {
+      setErrorMessage('Failed to load todos');
+      setTimeout(() => {
+        setErrorMessage('');
+      }, 3000);
+    }
+  }, []);
+
+  useEffect(() => {
+    if (editedTodoId) {
+      if (inputRef.current) {
+        inputRef.current.focus();
+      }
+    }
+  }, [editedTodoId]);
+
+  useEffect(() => {
+    handleLoadTodos();
+  }, []);
+
+  const hasCompletedTodos = todos.filter(todo => todo.completed).length > 0;
+  const hasActiveTodos = todos.filter(todo => !todo.completed).length > 0;
+
+  const filterVisibleTodos
+    = (filterList: FilterType, todoList: TodoType[]) => {
+      const filteredTodos = todoList;
+
+      return filteredTodos.filter(todo => {
+        switch (filterList) {
+          case FilterType.Completed:
+            return todo.completed;
+          case FilterType.Active:
+            return !todo.completed;
+          case FilterType.All:
+          default:
+            return todo;
+        }
+      });
+    };
+
+  const visibleTodos = useMemo(
+    () => filterVisibleTodos(filter, todos),
+    [todos, filterVisibleTodos],
+  );
+
+  const handleFilterChange = useCallback(
+    (newFilter: FilterType) => setFilter(newFilter),
+    [],
+  );
+
+  const handleCleanErrorMessage = useCallback(() => {
+    setErrorMessage('');
+  }, []);
+
+  const handleTodoTextChange = useCallback(
+    (event: React.ChangeEvent<HTMLInputElement>) => {
+      setTodoText(event.target.value);
+    }, [],
+  );
+
+  const handleNewTodoSubmit
+  = async (event: React.FormEvent<HTMLFormElement>) => {
+    event.preventDefault();
+
+    if (todoText.trim() === '') {
+      setErrorMessage('Title can\'t be empty');
+      setTimeout(() => {
+        setErrorMessage('');
+      }, 3000);
+
+      return;
+    }
+
+    const newTodo = {
+      id: 0,
+      userId: USER_ID,
+      title: todoText,
+      completed: false,
+    };
+
+    try {
+      setIsInputDisabled(true);
+      setTempTodo(newTodo);
+
+      const createdTodo = await addNewTodo(newTodo);
+
+      setTodos(prevTodos => [...prevTodos, createdTodo]);
+    } catch (error) {
+      setErrorMessage('Unable to add a todo');
```

Comment:
it would be better to set error in finally block

Reply:
Right now its implemented in a way that if errorMessage is empty string then if there is error i change it to adequate error message and if i move it to finally block i will have to create new state to hold information if there is error or not (set true of false in catch block) then i have to pass down two states as a props to children components. I can see it being worth doing for code clarity if working in a team but in this scenario ? If yes i will do it.